core/src/main/java/org/elasticsearch/index/fielddata/plain/AbstractGeoPointDVIndexFieldData.java

```diff
@@ -46,29 +50,41 @@ public final XFieldComparatorSource comparatorSource(@Nullable Object missingVal
         throw new IllegalArgumentException("can't sort on geo_point field without using specific sorting feature, like geo_distance");
     }
 
-    @Override
-    public AtomicGeoPointFieldData load(LeafReaderContext context) {
-        try {
-            return new GeoPointBinaryDVAtomicFieldData(DocValues.getBinary(context.reader(), fieldNames.indexName()));
-        } catch (IOException e) {
-            throw new IllegalStateException("Cannot load doc values", e);
+    /**
+     * Lucene 5.4 GeoPointFieldType
+     */
+    public static class GeoPointDVIndexFieldData extends AbstractGeoPointDVIndexFieldData {
+        final boolean indexCreatedBefore2x;
+
+        public GeoPointDVIndexFieldData(Index index, Names fieldNames, FieldDataType fieldDataType, final boolean indexCreatedBefore2x) {
+            super(index, fieldNames, fieldDataType);
+            this.indexCreatedBefore2x = indexCreatedBefore2x;
         }
-    }
 
-    @Override
-    public AtomicGeoPointFieldData loadDirect(LeafReaderContext context) throws Exception {
-        return load(context);
+        @Override
+        public AtomicGeoPointFieldData load(LeafReaderContext context) {
+            try {
+                return (indexCreatedBefore2x) ? new GeoPointLegacyDVAtomicFieldData(DocValues.getBinary(context.reader(), fieldNames.indexName())) :
+                        new GeoPointDVAtomicFieldData(DocValues.getSortedNumeric(context.reader(), fieldNames.indexName()));
+            } catch (IOException e) {
+                throw new IllegalStateException("Cannot load doc values", e);
```

Comment:
Is there anyway of outputting the name(s) of the fields we couldn't load doc values for here? Probably useful for debugging